### PR TITLE
Improvements for correct functioning of the Driver OCI8

### PR DESCRIPTION
**Pull request by desenvuno**
- Record count of a Select object;
- Problem solved in the definition of variable names in Oracle Select object when used subselects - limit and offset

### CODE REVIEW

File: library/Zend/Db/Adapter/Adapter.php [hunk 1]
```diff
@@ -333,8 +333,9 @@ protected function createPlatform($parameters)
                 // PDO is only supported driver for quoting values in this platform
                 return new Platform\SqlServer(($this->driver instanceof Driver\Pdo\Pdo) ? $this->driver : null);
             case 'Oracle':
+                $driver = ($this->driver instanceof Driver\Oci8\Oci8 || $this->driver instanceof Driver\Pdo\Pdo) ? $this->driver : null;
                 // oracle does not accept a driver as an option, no driver specific quoting available
```
Comment (Pittiplatsch) [inline library/Zend/Db/Adapter/Adapter.php]: Remove (now deprecated) comment

File: library/Zend/Db/Adapter/Driver/Oci8/Feature/RowCounter.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        
+        $sql = $statement->getSql();
+        if ($sql == '' || stripos(strtolower($sql), 'select') === false) {
+            return null;
+        }
+        $countSql = 'SELECT COUNT(*) as "count" FROM (' . $sql . ')';
+        $countStmt->prepare($countSql);        
+        $result = $countStmt->execute();
+        $countRow = $result->current();
+        unset($statement, $result);
```
Comment (Pittiplatsch) [inline library/Zend/Db/Adapter/Driver/Oci8/Feature/RowCounter.php]: What's the point of explicitly unsetting these vars here?

File: library/Zend/Db/Adapter/Driver/Oci8/Feature/RowCounter.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        return $countRow['count'];
+    }
+
+    /**
+     * @param $sql
```
Comment (Pittiplatsch) [inline library/Zend/Db/Adapter/Driver/Oci8/Feature/RowCounter.php]: `@param string $sql`

File: library/Zend/Db/Adapter/Driver/Oci8/Feature/RowCounter.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            return null;
+        }    
+        $countSql = 'SELECT COUNT(*) as "count" FROM (' . $sql . ')';        
+        $result = $this->driver->getConnection()->execute($countSql);
+        $countRow = $result->current();        
+        return $countRow['count'];
+    }
+
+    /**
+     * @param $context
```
Comment (Pittiplatsch) [inline library/Zend/Db/Adapter/Driver/Oci8/Feature/RowCounter.php]: `@param Statement|{???} $context`

File: library/Zend/Db/Adapter/Driver/Oci8/Oci8.php [hunk 5]
```diff
@@ -36,19 +43,44 @@ class Oci8 implements DriverInterface, Profiler\ProfilerAwareInterface
     protected $profiler = null;
 
     /**
+     * @var array
+     */
+    protected $options = array(
```
Comment (Pittiplatsch) [inline library/Zend/Db/Adapter/Driver/Oci8/Oci8.php]: Condense to one line

File: library/Zend/Db/Adapter/Driver/Oci8/Oci8.php [hunk 6]
```diff
@@ -130,6 +162,48 @@ public function getResultPrototype()
...
+    }
+
+    /**
+     * Setup the default features for Pdo
+     *
+     * @return Pdo
+     */
+    public function setupDefaultFeatures()
+    {        
+        $this->addFeature(null, new Feature\RowCounter);
```
Comment (Pittiplatsch) [inline library/Zend/Db/Adapter/Driver/Oci8/Oci8.php]: `new Feature\RowCounter()`

File: library/Zend/Db/Adapter/Driver/Oci8/Oci8.php [hunk 7]
```diff
@@ -130,6 +162,48 @@ public function getResultPrototype()
...
+        return $this;
+    }
+
+    /**
+     * Get feature
+     *
+     * @param $name
```
Comment (Pittiplatsch) [inline library/Zend/Db/Adapter/Driver/Oci8/Oci8.php]: `@param string $name`

File: library/Zend/Db/Adapter/Driver/Oci8/Result.php [hunk 8]
```diff
@@ -64,12 +64,13 @@ class Result implements Iterator, ResultInterface
      * @param resource $resource
      * @return Result
      */
-    public function initialize($resource /*, $generatedValue, $isBuffered = null*/)
+    public function initialize($resource, $generatedValue, $rowCount = null)
```
Comment (Pittiplatsch) [inline library/Zend/Db/Adapter/Driver/Oci8/Result.php]: *BC break*: Changed API

File: library/Zend/Db/Adapter/Platform/Oracle.php [hunk 9]
```diff
@@ -9,19 +10,46 @@
...
+
+    /** @var resource|\PDO */
+    protected $resource = null;
+
     /**
+     * @param null|\Zend\Db\Adapter\Driver\Oci8\Oci8|\Zend\Db\Adapter\Driver\Pdo\Pdo $driver $driver 
      * @param array $options
      */
-    public function __construct($options = array())
+    public function __construct($driver = null, $options = array())
```
Comment (Pittiplatsch) [inline library/Zend/Db/Adapter/Platform/Oracle.php]: BC break?

File: library/Zend/Db/Adapter/Platform/Oracle.php [hunk 10]
```diff
@@ -9,19 +10,46 @@
...
     {
-        if (isset($options['quote_identifiers'])
-            && ($options['quote_identifiers'] == false
-            || $options['quote_identifiers'] === 'false')
+        if (isset($options['quote_identifiers']) && ($options['quote_identifiers'] == false || $options['quote_identifiers'] === 'false')
```
Comment (Pittiplatsch) [inline library/Zend/Db/Adapter/Platform/Oracle.php]: Oversize line

File: library/Zend/Db/Adapter/Platform/Oracle.php [hunk 11]
```diff
@@ -9,19 +10,46 @@
...
+    }
+
+    /**
+     * @param \Zend\Db\Adapter\Driver\Pdo\Pdo||\PDO $driver
+     * @throws \Zend\Db\Adapter\Exception\InvalidArgumentException
+     * @return $this
+     */
+    public function setDriver($driver)
+    {
+        if ($driver instanceof Oci8\Oci8 || ($driver instanceof Pdo\Pdo && $driver->getDatabasePlatformName() == 'Oracle') || ($driver instanceof Pdo\Pdo && $driver->getDatabasePlatformName() == 'Sqlite') || ($driver instanceof \oci8) || ($driver instanceof \PDO && $driver->getAttribute(\PDO::ATTR_DRIVER_NAME) == 'oci')
```
Comment (Pittiplatsch) [inline library/Zend/Db/Adapter/Platform/Oracle.php]: Oversize line

File: library/Zend/Db/Adapter/Platform/Oracle.php [hunk 12]
```diff
@@ -49,11 +77,21 @@ public function quoteIdentifierChain($identifierChain)
...
+                return $this->resource->quote($value);
+            }
+        }
+
         trigger_error(
             'Attempting to quote a value in ' . __CLASS__ . ' without extension/driver support '
                 . 'can introduce security vulnerabilities in a production environment.'
         );
-        return '\'' . addcslashes(str_replace('\'', '\'\'', $value), "\x00\n\r\"\x1a") . '\'';
+        return '\'' . addcslashes(str_replace('\'', '\'\'', $value), "\x00\n\r\"\x1a") . '\'';
```
Comment (Pittiplatsch) [inline library/Zend/Db/Adapter/Platform/Oracle.php]: Trailing whitespace

File: tests/ZendTest/Db/Adapter/Platform/OracleTest.php [hunk 13]
```diff
@@ -83,8 +85,7 @@ public function testGetQuoteValueSymbol()
     public function testQuoteValueRaisesNoticeWithoutPlatformSupport()
     {
         $this->setExpectedException(
-            'PHPUnit_Framework_Error_Notice',
-            'Attempting to quote a value in Zend\Db\Adapter\Platform\Oracle without extension/driver support can introduce security vulnerabilities in a production environment'
+                'PHPUnit_Framework_Error_Notice', 'Attempting to quote a value in Zend\Db\Adapter\Platform\Oracle without extension/driver support can introduce security vulnerabilities in a production environment'
```
Comment (Pittiplatsch) [inline tests/ZendTest/Db/Adapter/Platform/OracleTest.php]: Wrong indentation?

File: library/Zend/Db/Adapter/Driver/Oci8/Oci8.php [hunk 14]
```diff
@@ -130,6 +162,48 @@ public function getResultPrototype()
     }
 
     /**
+     * Add feature
+     *
+     * @param string $name
+     * @param AbstractFeature $feature
+     * @return Pdo
```
Comment (samsonasik) [inline library/Zend/Db/Adapter/Driver/Oci8/Oci8.php]: @return self

File: library/Zend/Db/Adapter/Driver/Oci8/Oci8.php [hunk 15]
```diff
@@ -130,6 +162,48 @@ public function getResultPrototype()
...
+            $feature->setDriver($this);
+        }
+        $this->features[$name] = $feature;
+        return $this;
+    }
+
+    /**
+     * Setup the default features for Pdo
+     *
+     * @return Pdo
```
Comment (samsonasik) [inline library/Zend/Db/Adapter/Driver/Oci8/Oci8.php]: @return self

File: library/Zend/Db/Adapter/Driver/Oci8/Statement.php [hunk 16]
```diff
@@ -308,9 +307,24 @@ protected function bindParametersFromContainer()
...
-            }
-
+            }            
             oci_bind_by_name($this->resource, $name, $value, $maxLength, $type);
         }
     }
+    
+    /**
+     * Perform a deep clone
+     * @return Statement A cloned statement
```
Comment (samsonasik) [inline library/Zend/Db/Adapter/Driver/Oci8/Statement.php]: no return ?

File: library/Zend/Db/Adapter/Platform/Oracle.php [hunk 17]
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
Comment (samsonasik) [inline library/Zend/Db/Adapter/Platform/Oracle.php]: remove this blank line

**Comment by samsonasik** [comment]
if you use php-cs-fixer to fix the files, please do with --config-file like this : 
```shell
php ./vendor/bin/php-cs-fixer fix  path/to/dir/or/file -v --diff --dry-run --config-file=.php_cs
```

File: library/Zend/Db/Adapter/Driver/Oci8/Oci8.php (hunk 15 above)
Comment (Maks3w) [inline library/Zend/Db/Adapter/Driver/Oci8/Oci8.php]: @samsonasik `@return $this` is recognized by phpDocumentor 2 and has a better meaning

File: library/Zend/Db/Adapter/Platform/Oracle.php (hunk 9 above)
Comment (romulobusatto) [inline library/Zend/Db/Adapter/Platform/Oracle.php]: I don't understand. What do with that?

Comment (samsonasik) [inline library/Zend/Db/Adapter/Platform/Oracle.php]: changing parameter order will break compatibility (major change)

**Comment by romulobusatto** [comment]
After execution of the tests, submitted one error of assertion in php 5.6, but is resource not modified by us. Can help us?

File: library/Zend/Db/Adapter/Driver/Oci8/Oci8.php (hunk 15 above)
Comment (Martin-P) [inline library/Zend/Db/Adapter/Driver/Oci8/Oci8.php]: Just as a reference: https://github.com/zendframework/zf2/pull/4436#issuecomment-17538536

**Comment by FraGoTe** [comment]
Hi, it would be great to have a function to fetch cursors like this:

https://github.com/FraGoTe/cursor-oracle-zend-framework-adapter/blob/master/Zend_Db_Adapter_Cds.php

**Comment by desenvuno** [comment]
Hi @weierophinney. We are following this PR, can you tell us If there is anything we need to do?

**Comment by samsonasik** [comment]
could you send to zend-db repo instead https://github.com/zendframework/zend-db ?

**Comment by desenvuno** [comment]
Created @samsonasik, here https://github.com/zendframework/zend-db/pull/9

**Pull request closed without merging**